pkg/log/context.go

```diff
@@ -23,92 +23,145 @@ import (
+	Output string `description:"output for the logs. Allowed values - /dev/stdout, /dev/stderr, ginkgowriter" json:"output,omitempty"`
 }
 
 // DefaultSettings returns default values for Log settings
 func DefaultSettings() *Settings {
 	return &Settings{
 		Level:  "error",
 		Format: "text",
-		Output: os.Stdout,
+		Output: os.Stdout.Name(),
 	}
 }
 
 // Validate validates the logging settings
 func (s *Settings) Validate() error {
 	if _, err := logrus.ParseLevel(s.Level); err != nil {
-		return fmt.Errorf("validate Settings: %s", err)
+		return fmt.Errorf("validate Settings: log level %s is invalid: %s", s.Level, err)
 	}
+
 	if len(s.Format) == 0 {
-		return fmt.Errorf("validate Settings: LogFormat missing")
+		return fmt.Errorf("validate Settings: log format missing")
 	}
-	if s.Output == nil {
-		return fmt.Errorf("validate Settings: LogOutput missing")
+
+	if _, ok := supportedFormatters[s.Format]; !ok {
+		return fmt.Errorf("validate Settings: log format %s is invalid", s.Format)
+	}
+
+	if len(s.Output) == 0 {
+		return fmt.Errorf("validate Settings: log output missing")
 	}
+
+	if _, ok := supportedOutputs[s.Output]; !ok {
+		return fmt.Errorf("validate Settings: log output %s is invalid", s.Output)
+	}
+
 	return nil
 }
 
 // Configure creates a new context with a logger using the provided settings.
-func Configure(ctx context.Context, settings *Settings) context.Context {
-	once.Do(func() {
-		level, err := logrus.ParseLevel(settings.Level)
-		if err != nil {
-			panic(fmt.Sprintf("Could not parse log level configuration: %s", err))
-		}
-		formatter, ok := supportedFormatters[settings.Format]
-		if !ok {
-			panic(fmt.Sprintf("Invalid log format: %s", settings.Format))
-		}
-		logger := &logrus.Logger{
-			Formatter: formatter,
-			Level:     level,
-			Out:       settings.Output,
-			Hooks:     make(logrus.LevelHooks),
-		}
-		hook := filename.NewHook()
-		hook.Field = FieldComponentName
-		logger.AddHook(hook)
-		defaultEntry = logrus.NewEntry(logger)
-		defaultEntry = defaultEntry.WithField(FieldCorrelationID, "-")
-	})
-	return ContextWithLogger(ctx, defaultEntry)
+func Configure(ctx context.Context, settings *Settings) (context.Context, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	currentSettings = settings
+	level, err := logrus.ParseLevel(settings.Level)
+	if err != nil {
+		return nil, fmt.Errorf("invalid log level: %s", err)
+	}
+	formatter, ok := supportedFormatters[settings.Format]
+	if !ok {
+		return nil, fmt.Errorf("invalid log format: %s", settings.Format)
+	}
+
+	output, ok := supportedOutputs[settings.Output]
+	if !ok {
+		return nil, fmt.Errorf("invalid output: %s", settings.Output)
+	}
+
+	entry := ctx.Value(logKey{})
+	if entry == nil {
+		entry = defaultEntry
+	} else {
+		defaultEntry.Logger.SetOutput(output)
+		defaultEntry.Logger.SetFormatter(formatter)
+		defaultEntry.Logger.SetLevel(level)
+	}
+	e := copyEntry(entry.(*logrus.Entry))
```

Comment:
This will change the log level of the configuration request. Do we want that? e.g. you log on error before the controller and log on debug after it.

Reply:
I am not sure - probably it would be easier to understand if the log level for the current request is kept the same - we could do `	_, err = log.Configure(ctx, &loggingConfig)
` and skip assigning the new ctx in the log controller